File: understand-anything-plugin/src/onboard-builder.ts
import type { KnowledgeGraph } from "@understand-anything/core";

/**
 * Generate a structured onboarding guide from the knowledge graph.
 * Output is standalone markdown suitable for a README, wiki, or docs.
 */
export function buildOnboardingGuide(graph: KnowledgeGraph): string {
  const { project, nodes, edges, layers } = graph;
  const lines: string[] = [];

  // --- Project Overview ---
  lines.push(`# ${project.name}`);
  lines.push("");
  lines.push(`> ${project.description}`);
  lines.push("");
  lines.push(`| | |`);
  lines.push(`|---|---|`);
  lines.push(`| **Languages** | ${project.languages.join(", ")} |`);
  lines.push(`| **Frameworks** | ${project.frameworks.join(", ")} |`);
  lines.push(`| **Components** | ${nodes.length} nodes, ${edges.length} relationships |`);
  lines.push(`| **Last Analyzed** | ${project.analyzedAt} |`);
  lines.push("");

  // --- Architecture ---
  if (layers.length > 0) {
    lines.push("## Architecture");
    lines.push("");
    lines.push("The project is organized into the following layers:");
    lines.push("");

    for (const layer of layers) {
      const memberNames = layer.nodeIds
        .map((id) => nodes.find((n) => n.id === id)?.name)
        .filter(Boolean);
      lines.push(`### ${layer.name}`);
      lines.push("");
      lines.push(layer.description);
      lines.push("");
      if (memberNames.length > 0) {
        lines.push(`Key components: ${memberNames.join(", ")}`);
        lines.push("");
      }
    }
  }

  // --- Key Concepts ---
  const conceptNodes = nodes.filter((n) => n.type === "concept");
  if (conceptNodes.length > 0) {
    lines.push("## Key Concepts");
    lines.push("");
    lines.push("Important architectural and domain concepts to understand:");
    lines.push("");
    for (const concept of conceptNodes) {
      lines.push(`### ${concept.name}`);
      lines.push("");
      lines.push(concept.summary);
      lines.push("");
    }
  }

  // --- Getting Started ---
  // Derive a deterministic reading order from the dependency graph: the most
  // depended-on files (highest fan-in via imports/depends_on) are foundational
  // and worth reading first.
  const dependencyEdgeTypes = new Set(["imports", "depends_on"]);
  const fanIn = new Map<string, number>();
  for (const edge of edges) {
    if (!dependencyEdgeTypes.has(edge.type)) continue;
    // For backward edges the dependency points the other way.
    const depended = edge.direction === "backward" ? edge.source : edge.target;
    fanIn.set(depended, (fanIn.get(depended) ?? 0) + 1);
  }

  const foundationalFiles = nodes
    .filter((n) => n.type === "file" && n.filePath && (fanIn.get(n.id) ?? 0) > 0)
    .sort((a, b) => {
      const diff = (fanIn.get(b.id) ?? 0) - (fanIn.get(a.id) ?? 0);
      return diff !== 0 ? diff : a.filePath!.localeCompare(b.filePath!);
    })
    .slice(0, 8);

  if (foundationalFiles.length > 0 || layers.length > 0) {
    lines.push("## Getting Started");
    lines.push("");

    if (foundationalFiles.length > 0) {
      lines.push(
        "Start with the most depended-on files — these are the foundations the rest of the code builds on:",
      );
      lines.push("");
      for (const node of foundationalFiles) {
        const count = fanIn.get(node.id) ?? 0;
        lines.push(
          `- \`${node.filePath}\` — ${node.summary} _(referenced by ${count} component${count === 1 ? "" : "s"})_`,
        );
      }
      lines.push("");
    }

    if (layers.length > 0) {
      lines.push(
        "Then explore the codebase layer by layer, in the order listed under **Architecture** above.",
      );
      lines.push("");
    }
  }

  // --- File Map ---
  const fileNodes = nodes.filter((n) => n.type === "file" && n.filePath);
  if (fileNodes.length > 0) {
    lines.push("## File Map");
    lines.push("");
    lines.push("| File | Purpose | Complexity |");
    lines.push("|------|---------|------------|");
    for (const node of fileNodes) {
      lines.push(`| \`${node.filePath}\` | ${node.summary} | ${node.complexity} |`);
    }
    lines.push("");
  }

  // --- Complexity Hotspots ---
  const complexNodes = nodes.filter((n) => n.complexity === "complex");
  if (complexNodes.length > 0) {
    lines.push("## Complexity Hotspots");
    lines.push("");
    lines.push("These components are the most complex and deserve extra attention:");
    lines.push("");
    for (const node of complexNodes) {
      lines.push(`- **${node.name}** (${node.type}): ${node.summary}`);
    }
    lines.push("");
  }

  // --- Footer ---
  lines.push("---");
  lines.push("");
  lines.push(`*Generated from knowledge graph v${graph.version}*`);
  lines.push("");

  return lines.join("\n");
}

File: understand-anything-plugin/packages/core/skill_builders/onboard_builder.py
"""Generate a standalone onboarding guide — port of ``onboard-builder.ts``."""

from __future__ import annotations

from understand_core import KnowledgeGraph


def build_onboarding_guide(graph: KnowledgeGraph) -> str:
    """Produce a markdown onboarding guide (README/wiki-ready) from the graph."""
    project, nodes, edges, layers = (
        graph.project,
        graph.nodes,
        graph.edges,
        graph.layers,
    )
    lines: list[str] = []

    # --- Project Overview ---
    lines.append(f"# {project.name}")
    lines.append("")
    lines.append(f"> {project.description}")
    lines.append("")
    lines.append("| | |")
    lines.append("|---|---|")
    lines.append(f"| **Languages** | {', '.join(project.languages)} |")
    lines.append(f"| **Frameworks** | {', '.join(project.frameworks)} |")
    lines.append(
        f"| **Components** | {len(nodes)} nodes, {len(edges)} relationships |"
    )
    lines.append(f"| **Last Analyzed** | {project.analyzedAt} |")
    lines.append("")

    # --- Architecture ---
    if layers:
        lines.append("## Architecture")
        lines.append("")
        lines.append("The project is organized into the following layers:")
        lines.append("")

        name_by_id = {n.id: n.name for n in nodes}
        for layer in layers:
            member_names = [
                name_by_id[node_id]
                for node_id in layer.nodeIds
                if name_by_id.get(node_id)
            ]
            lines.append(f"### {layer.name}")
            lines.append("")
            lines.append(layer.description)
            lines.append("")
            if member_names:
                lines.append(f"Key components: {', '.join(member_names)}")
                lines.append("")

    # --- Key Concepts ---
    concept_nodes = [n for n in nodes if n.type == "concept"]
    if concept_nodes:
        lines.append("## Key Concepts")
        lines.append("")
        lines.append("Important architectural and domain concepts to understand:")
        lines.append("")
        for concept in concept_nodes:
            lines.append(f"### {concept.name}")
            lines.append("")
            lines.append(concept.summary)
            lines.append("")

    # --- Getting Started ---
    # Derive a deterministic reading order from the dependency graph: the most
    # depended-on files (highest fan-in via imports/depends_on) are foundational
    # and worth reading first.
    dependency_edge_types = {"imports", "depends_on"}
    fan_in: dict[str, int] = {}
    for edge in edges:
        if edge.type not in dependency_edge_types:
            continue
        # For backward edges the dependency points the other way.
        direction = getattr(edge, "direction", "forward")
        depended = edge.source if direction == "backward" else edge.target
        fan_in[depended] = fan_in.get(depended, 0) + 1

    foundational_candidates = [
        n
        for n in nodes
        if n.type == "file" and n.filePath and fan_in.get(n.id, 0) > 0
    ]
    foundational_candidates.sort(
        key=lambda n: (-fan_in.get(n.id, 0), n.filePath or "")
    )
    foundational_files = foundational_candidates[:8]

    if foundational_files or layers:
        lines.append("## Getting Started")
        lines.append("")

        if foundational_files:
            lines.append(
                "Start with the most depended-on files — these are the foundations the rest of the code builds on:"
            )
            lines.append("")
            for node in foundational_files:
                count = fan_in.get(node.id, 0)
                plural = "" if count == 1 else "s"
                lines.append(
                    f"- `{node.filePath}` — {node.summary} _(referenced by {count} component{plural})_"
                )
            lines.append("")

        if layers:
            lines.append(
                "Then explore the codebase layer by layer, in the order listed under **Architecture** above."
            )
            lines.append("")

    # --- File Map ---
    file_nodes = [n for n in nodes if n.type == "file" and n.filePath]
    if file_nodes:
        lines.append("## File Map")
        lines.append("")
        lines.append("| File | Purpose | Complexity |")
        lines.append("|------|---------|------------|")
        for node in file_nodes:
            lines.append(
                f"| `{node.filePath}` | {node.summary} | {node.complexity} |"
            )
        lines.append("")

    # --- Complexity Hotspots ---
    complex_nodes = [n for n in nodes if n.complexity == "complex"]
    if complex_nodes:
        lines.append("## Complexity Hotspots")
        lines.append("")
        lines.append(
            "These components are the most complex and deserve extra attention:"
        )
        lines.append("")
        for node in complex_nodes:
            lines.append(f"- **{node.name}** ({node.type}): {node.summary}")
        lines.append("")

    # --- Footer ---
    lines.append("---")
    lines.append("")
    lines.append(f"*Generated from knowledge graph v{graph.version}*")
    lines.append("")

    return "\n".join(lines)
